### PR TITLE
feat: top-level api cleanup, xrp-paychan, simplify prefunding 

### PR DESCRIPTION
**Pull request by kincaidoneil**
### Top-level API
- `deposit` -- move money from layer 1 into layer 2 for the provided uplink
- `withdraw` -- move money from layer 2 to layer 1 for the provided uplink
-  `remove` -- permanently delete the provided uplink
- `add` -- add a new uplink using the provided credential (previously `configure`)
- `disconnect` -- gracefully disconnect the entire API
- Much cleaner internal interfaces and separation of credentials/engine/uplink code (but still more work to be done)

### Xrp Refactoring
- Migrate to ilp-plugin-xrp-paychan from asym client/server (more robust/secure codebase, and allows creating ad-hoc channels without a very particular handshake on connect)
- Introduces some bugs/less reliable as the uplink hasn't been tested as thoroughly
- Requires v23 connector refactor to be running on server to run xrp-paychan, but on the plus side, much simpler stack since there's more symmetry between the server & client
- Minor changes to xrp-paychan are required (also requires sitting PR on plugin-btp)

### Prefunding
- Instead of always maintaining a prefund of _x_ with the connector, the new behavior only prefunds directly before sending a packet
- This slightly increases latency for the first packet, but otherwise is nearly equally as performant
- Simplifies code: no longer necessary to stream money off the connector on disconnect
- Simplifies accounting/UX: prefunding is no longer included in the `balance$` observable since prefunds only occur during streaming, which makes the front-end/user-facing UX **much** simpler (Lnd prefunding on connect presented the question, do you have to authorize the prefund when the user first adds the uplink? What if the prefund amount changes? etc. This resolves those questions)

### CODE REVIEW

File: src/engine.ts [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+   */
+  remoteConnectors: {
+    readonly [name: string]: (token: string) => string
+  }
+}
+
+export type SettlementEngines = LndSettlementEngine | XrpPaychanSettlementEngine
+
+// TODO Add "closeEngine" to specific settlement modules
+export const closeEngine = (settler: SettlementEngines) => {
```
Comment (karzak) [inline src/engine.ts]: What would 'closeEngine' be responsible for?

File: src/credential.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+export type ValidatedCredentials = (
+  | ValidatedLndCredential
+  | ValidatedXrpSecret) & {
+  settlerType: SettlementEngineType
+}
+
+export type ReadyCredentials = (ReadyLndCredential | ReadyXrpCredential) & {
+  settlerType: SettlementEngineType
+}
+
```
Comment (karzak) [inline src/credential.ts]: We do this pattern a lot, and as we add more settlers, this could get quite verbose. Not sure if it's bad, just something I noticed.

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    await Promise.all(
+      Object.values(state.settlers)
+        .filter((a): a is SettlementEngines => !!a) // TODO !
+        .map(closeEngine)
+    )
+  }
+
+  // TODO Should disconnecting the API prevent other operations from occuring?
+
+  return {
```
Comment (karzak) [inline src/index.ts]: Seems pretty clean and idiomatic 👍

File: src/credential.ts (hunk 2 above)
Comment (karzak) [reply src/credential.ts]: Should be referring to the `setupCredential` code, not ReadyCredentials.

Comment (rhuairahrighairidh) [reply src/credential.ts]: Also makes unit testing a bit harder. You can't pass a mocked object to a function if the arg type is fixed to being one of xrp,lnd,etc.

File: src/engine.ts (hunk 1 above)
Comment (kincaidoneil) [reply src/engine.ts]: Probably not much right now, e.g. just closing the eth node connection, or disconnecting from Ripple. If we ran a light client (or something with more overhead) that was shared between multiple uplinks, it'd provide a nice hook to gracefully shut it down.

**Comment by rhuairahrighairidh** [comment]
I got into a bit of a muddle end of last sprint understanding some of the concepts. So here's some of the things I was confused about, if it's useful.

 - naming confused me a bit
    - `validatedCredentials` are actually the unvalidated credentials?
    - engines are sometimes referred to as settlers
    - `settlementEngine` sounds like a thing that settles money rather than asset and connector uri data?
    - lnd credentials contain objects for talking to the lnd node rather than just static credentials
 - logic was also a bit hard to follow
    - not sure what functions have side effects/mutate state and what don’t
    - wasn't sure when some function does something simple like format a string vs does something consequential like move money / register with connector
 - found the concept of credentials, engines, uplinks and their relationships a bit hard to understand.
